### PR TITLE
schema: add SR-IOV basic schema

### PR DESCRIPTION
**Pull request by ffmancera**
Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

### CODE REVIEW

**Comment by ffmancera** [comment]
This is a proposal for a basic SR-IOV schema with this we will be able to set and get the total number of VFs.

File: libnmstate/schema.py [hunk 1]
```diff
@@ -200,6 +200,14 @@ class Ethernet(object):
     FULL_DUPLEX = 'full'
     HALF_DUPLEX = 'half'
 
+    SR_IOV = 'sr-iov'
+
+
+class SRIOV(object):
+    CONFIG_SUBTREE = 'sr-iov'
```
Comment (tyll) [inline libnmstate/schema.py]: You can re-use the `Ethernet.SR_IOV` constant here. Can you please also explain in the commit message why it is a subtree of Ethernet?

File: tests/lib/schema_validation_test.py [hunk 2]
```diff
@@ -170,6 +170,10 @@ def test_valid_without_auto_neg_and_missing_speed(self, default_data):
 
         libnmstate.validator.validate(default_data)
 
+    def test_valid_with_sriov_total_vfs(self, default_data):
+        default_data[Interface.KEY][0].update({'sr-iov': {'total-vfs': 5}})
```
Comment (tyll) [inline tests/lib/schema_validation_test.py]: Please use the constants here for `sr-iov` and `total-vfs`.

**Comment by tyll** [comment]
We probably didn't do it in the past but IMHO it would make sense to add a simple code snippet to Nmstate that will raise an exception that SR-IOV is not yet supported if it is specified in a state to keep the code consistent. Otherwise the schema implies to an outsider that it is supported but Nmstate will silently fail.

File: libnmstate/schema.py (hunk 1 above)
Comment (ffmancera) [reply libnmstate/schema.py]: Done! The explanation added is: `SR-IOV is a subtree of ethernet because it contains the PF information which must be an ethernet interface.`

File: tests/lib/schema_validation_test.py (hunk 2 above)
Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: Done! Thanks!

**Comment by ffmancera** [comment]
> We probably didn't do it in the past but IMHO it would make sense to add a simple code snippet to Nmstate that will raise an exception that SR-IOV is not yet supported if it is specified in a state to keep the code consistent. Otherwise the schema implies to an outsider that it is supported but Nmstate will silently fail.

I've added a `nm/sriov.py` placeholder which raises a `NmstateNotImplementedError` exception if nmstate finds `SRIOV.CONFIG_SUBTREE` in a desired interface state.

File: libnmstate/nm/sriov.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from libnmstate.error import NmstateNotImplementedError
+from libnmstate.schema import SRIOV
+
+
+def create_setting(iface_state):
+    setting = iface_state.get(SRIOV.CONFIG_SUBTREE)
+    if setting:
+        raise NmstateNotImplementedError
+
+    return setting
```
Comment (tyll) [inline libnmstate/nm/sriov.py]: While the code works as expected, using the variable `setting` for both the config and the return value is a little bit misleading because it would have to return some NM object and not a a changed config subtree. Therefore `return None` would be better and maybe even using `config` instead of `setting` for the variable that contains the config. Or just avoid the variables,

```
if iface_state.get(SRIOV.CONFIG_SUBTREE):
    raise NmstateNotImplementedError

return None
```

Ideally you would also add a small integration tests that ensures that `create_setting()` raises the exception when the config is used in  state.

File: libnmstate/schemas/operational-state.yaml [hunk 4]
```diff
@@ -163,6 +163,12 @@ definitions:
           minimum: 0
         flow-control:
           type: boolean
+        sr-iov:
+          type: object
+          properties:
+            total-vfs:
```
Comment (maiqueb) [inline libnmstate/schemas/operational-state.yaml]: I think you should also set the max value of total-vfs to 256.

**Comment by ffmancera** [comment]
> Since you've bound the variable in the schema (min=0) - and I'm requesting you to also set a max - I think you should add some negative testing for those scenarios.
> 
> A question: if I understand correctly, your input for this task is bug [0]. Why are you only defining the max number of virtual functions ? (I'm 100% sure I'm missing some somethings :) )
> 
> [0] - https://bugzilla.redhat.com/show_bug.cgi?id=1695116

That seems fine, thanks! We have decided to implement first the basics and then extend that. You can check sprint planning for this :-)

File: tests/lib/schema_validation_test.py [hunk 5]
```diff
@@ -170,6 +171,34 @@ def test_valid_without_auto_neg_and_missing_speed(self, default_data):
...
+    def test_negative_total_vfs_is_invalid(self, default_data):
+        default_data[Interface.KEY][0].update(
+            {SRIOV.CONFIG_SUBTREE: {SRIOV.TOTAL_VFS: -3}}
+        )
+        libnmstate.validator.validate(default_data)
+
+        with pytest.raises(js.ValidationError) as err:
+            libnmstate.validator.validate(default_data)
+
+        assert '-3 is lower than the minimum of 0' in err.value.args[0]
```
Comment (maiqueb) [inline tests/lib/schema_validation_test.py]: wondering if you can't use the pytest.raises(... **match**) keyword here instead of this (IMHO) ugly assert.

100 % optional, I just think it has potential to look better.

File: tests/lib/schema_validation_test.py [hunk 6]
```diff
@@ -170,6 +171,34 @@ def test_valid_without_auto_neg_and_missing_speed(self, default_data):
...
+    def test_over_maximum_total_vfs_is_invalid(self, default_data):
+        default_data[Interface.KEY][0].update(
+            {SRIOV.CONFIG_SUBTREE: {SRIOV.TOTAL_VFS: 300}}
+        )
+        libnmstate.validator.validate(default_data)
+
+        with pytest.raises(js.ValidationError) as err:
+            libnmstate.validator.validate(default_data)
+
+        assert '300 is greater than the maximum of 256' in err.value.args[0]
```
Comment (maiqueb) [inline tests/lib/schema_validation_test.py]: ditto.

File: tests/integration/sriov_test.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import libnmstate
+from libnmstate.error import NmstateNotImplementedError
+from libnmstate.schema import Ethernet
+from libnmstate.schema import Interface
+from libnmstate.schema import InterfaceState
+from libnmstate.schema import InterfaceType
+from libnmstate.schema import SRIOV
+
+
+@pytest.mark.xfail(
```
Comment (maiqueb) [inline tests/integration/sriov_test.py]: Hm! Neat. Does this throw you a warning resembling this one ? "UserWarning: Network state after test run does not match network state before test run:"

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (ffmancera) [reply libnmstate/schemas/operational-state.yaml]: Done! Thanks!

File: tests/lib/schema_validation_test.py (hunk 5 above)
Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: It looks much better. Thanks!

File: tests/lib/schema_validation_test.py (hunk 6 above)
Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: Same here. Please look above

File: tests/integration/sriov_test.py (hunk 7 above)
Comment (ffmancera) [reply tests/integration/sriov_test.py]: No. But I think this warning is not related to your changes. You can check it using `pytest-args='-k <your_test_name>'` when executing run-test.sh

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (tyll) [reply libnmstate/schemas/operational-state.yaml]: > I think you should also set the max value of total-vfs to 256.

Where does this number come from? I was wondering if there was an upper limit as well but it seemed to depend on the actual hardware. Also `256` is a strange limit, if it corresponds to an unsigned byte, the maximum would be `255` since the minimum is `0` - or does `0` indicate that there is no support and there could be `256` different VFs with IDs `0` to `255`?

File: libnmstate/nm/sriov.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+from libnmstate.error import NmstateNotImplementedError
+from libnmstate.schema import Ethernet
+from libnmstate.schema import SRIOV
+
+
+def create_setting(iface_state):
+    setting = None
+    ethernet_setting = iface_state.get(Ethernet.CONFIG_SUBTREE)
+    if ethernet_setting:
```
Comment (tyll) [inline libnmstate/nm/sriov.py]: You can simplify this as

```
setting = None
sriov_config = iface_state.get(Ethernet.CONFIG_SUBTREE, {}).get(SRIOV.CONFIG_SUBTREE)
if sriov_config:
    raise NmstateNotImplementedError

return setting
```
This avoids the multiple ifs. Hope this will please the linter.

File: tests/lib/schema_validation_test.py [hunk 9]
```diff
@@ -170,6 +171,30 @@ def test_valid_without_auto_neg_and_missing_speed(self, default_data):
 
         libnmstate.validator.validate(default_data)
 
+    def test_valid_with_sriov_total_vfs(self, default_data):
```
Comment (tyll) [inline tests/lib/schema_validation_test.py]: Personally, I like testing the edge cases explicitly, which would be -1, 0, 256, 257 and maybe few extra values beside these. Since there are two categories, valid and invalid, it makes in my opinion sense to use `@pytest.parametrize` and one test for valid values and one for invalid values. What do you think?

File: tests/integration/sriov_test.py (hunk 7 above)
Comment (tyll) [reply tests/integration/sriov_test.py]: @maiqueb it is unlikely that this test will expose this because `eth1` is cleaned up a lot of times. I guess for virtual interfaces, it will happen less likely when #593 is merged.

File: tests/integration/sriov_test.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    raises=NmstateNotImplementedError,
+    reason='SR-IOV is not supported yet',
+    strict=True,
+)
+def test_sriov_not_implemented():
+    desired_state = {
+        Interface.KEY: [
+            {
+                Interface.NAME: 'eth1',
```
Comment (tyll) [inline tests/integration/sriov_test.py]: Please make this a constant, maybe `SRIOV_NIC`. Even when it is implemented, this will only work on special hardware and not in the CI.

Comment (ffmancera) [reply tests/integration/sriov_test.py]: Yes, in the implementation PR we will need to check if we have SR-IOV support or not.

File: tests/lib/schema_validation_test.py (hunk 9 above)
Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: That seems better. Will do, thanks!

File: libnmstate/nm/sriov.py (hunk 8 above)
Comment (ffmancera) [reply libnmstate/nm/sriov.py]: Yes, that is fine. Thanks!

File: tests/lib/schema_validation_test.py (hunk 9 above)
Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: Done!

File: tests/integration/sriov_test.py (hunk 7 above)
Comment (maiqueb) [reply tests/integration/sriov_test.py]: tks @tyll .

File: tests/integration/sriov_test.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    Ethernet.CONFIG_SUBTREE: {SRIOV.CONFIG_SUBTREE: {SRIOV.TOTAL_VFS: 2}},
+}
+
+
+@pytest.mark.xfail(
+    raises=NmstateNotImplementedError,
+    reason='SR-IOV is not supported yet',
+    strict=True,
+)
+def test_sriov_not_implemented():
```
Comment (tyll) [inline tests/integration/sriov_test.py]: This should probably use the `eth1_up` fixture since the state uses `eth1`.

File: tests/integration/sriov_test.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import libnmstate
+from libnmstate.error import NmstateNotImplementedError
+from libnmstate.schema import Ethernet
+from libnmstate.schema import Interface
+from libnmstate.schema import InterfaceState
+from libnmstate.schema import InterfaceType
+from libnmstate.schema import SRIOV
+
+
+SRIOV_NIC = {
```
Comment (tyll) [inline tests/integration/sriov_test.py]: Initially I meant to use `SRIOV_NIC = "eth1"`. If you are using it for the full state, it should be something like SRIOV_CONFIG.

File: tests/integration/sriov_test.py (hunk 11 above)
Comment (ffmancera) [reply tests/integration/sriov_test.py]: Done!

File: tests/integration/sriov_test.py (hunk 12 above)
Comment (ffmancera) [reply tests/integration/sriov_test.py]: Done! Thanks!

**Comment by tyll** [comment]
It seems that the integration test is not working as expected (it is not failing): https://travis-ci.com/nmstate/nmstate/jobs/260153026

**Comment by ffmancera** [comment]
> It seems that the integration test is not working as expected (it is not failing): https://travis-ci.com/nmstate/nmstate/jobs/260153026

Yes, just a typo. Fixing it.

File: tests/lib/schema_validation_test.py [hunk 13]
```diff
@@ -170,6 +181,30 @@ def test_valid_without_auto_neg_and_missing_speed(self, default_data):
...
+    ):
+        default_data[Interface.KEY][0].update(
+            {
+                Interface.TYPE: InterfaceType.ETHERNET,
+                SRIOV.CONFIG_SUBTREE: {SRIOV.TOTAL_VFS: invalid_values},
+            }
+        )
+
+        with pytest.raises(js.ValidationError):
+            libnmstate.validator.validate(default_data)
```
Comment (maiqueb) [inline tests/lib/schema_validation_test.py]: shouldn't you verify the error msg as well ? What I had suggested in a previous comment was for you to use the *match* keyword instead, no for your to stop checking that.

I meant something like:
```
with pytest.raises(ValueError, match=r".* 123 .*"):
    myfunc()
```

From [0] I get the understanding this is pytest's way of asserting on py3.

[0] - https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: That seems fine. Done! Thanks :-)

File: libnmstate/schemas/operational-state.yaml [hunk 14]
```diff
@@ -163,6 +163,13 @@ definitions:
...
+              type: integer
+              minimum: 0
+              maximum: 256
```
Comment (maiqueb) [inline libnmstate/schemas/operational-state.yaml]: Should this allow 256? I think it shouldn't.

File: tests/lib/schema_validation_test.py [hunk 15]
```diff
@@ -170,6 +181,30 @@ def test_valid_without_auto_neg_and_missing_speed(self, default_data):
...
+        self, default_data, invalid_values
+    ):
+        default_data[Interface.KEY][0].update(
+            {
+                Interface.TYPE: InterfaceType.ETHERNET,
+                SRIOV.CONFIG_SUBTREE: {SRIOV.TOTAL_VFS: invalid_values},
+            }
+        )
+
+        with pytest.raises(js.ValidationError, match=str(invalid_values)):
```
Comment (maiqueb) [inline tests/lib/schema_validation_test.py]: not sure it warrants the trouble, but would be better to have it w/ 

```python
invalid_value = 300
max_or_min = 'maximum' if invalid_value > 255 else 'minimum'
comparator = 'greater' if invalid_value > 255 else 'smaller'
with pytest.raises(js.ValidationError, match=f'{invalid_value} is {comparator} than the allowed {max_or_min} of 256'):
    ...
```
Feel free to disregard.

File: libnmstate/schemas/operational-state.yaml (hunk 14 above)
Comment (tyll) [reply libnmstate/schemas/operational-state.yaml]: The number came from you: https://github.com/nmstate/nmstate/pull/592#discussion_r349574736
Now that you mention it again, I remember that I asked you about it as well:
https://github.com/nmstate/nmstate/pull/592#discussion_r349795251

File: tests/lib/schema_validation_test.py (hunk 15 above)
Comment (tyll) [reply tests/lib/schema_validation_test.py]: This seems to be testing too many internals of jsonschema that Nmstate does not care about (does it @EdDev ?)

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (maiqueb) [reply libnmstate/schemas/operational-state.yaml]: This is the only reference I could find of the 256 number.
https://www.itprotoday.com/compute-engines/q-what-sr-iov

Even that doc refers it's HW specific, so nmstate shouldn't meddle.

Sorry for misleading you @ffmancera. I think you should remove the 'maximum' .

Comment (ffmancera) [reply libnmstate/schemas/operational-state.yaml]: No problem about that. I am going to remove the `maximum`. Thanks!

File: tests/lib/schema_validation_test.py (hunk 15 above)
Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: If we are removing the `maximum` limit then we can simplify this a lot. Thanks!

File: libnmstate/schema.py [hunk 16]
```diff
@@ -200,6 +200,13 @@ class Ethernet(object):
     FULL_DUPLEX = 'full'
     HALF_DUPLEX = 'half'
 
+    SR_IOV = 'sr-iov'
+
+    class SRIOV(object):
+        CONFIG_SUBTREE = 'sr-iov'
```
Comment (tyll) [inline libnmstate/schema.py]: AFAIU the LinuxBridge example, this constant should be a member of the Ethernet class.

Comment (ffmancera) [reply libnmstate/schema.py]: Right, done! Thanks

File: tests/lib/schema_validation_test.py [hunk 17]
```diff
@@ -145,6 +147,15 @@ def test_invalid_type(self, default_data):
...
 
 class TestIfaceTypeEthernet(object):
+
+    parametrize_sriov_valid_total_vfs = pytest.mark.parametrize(
+        'valid_values', [0, 150, 256]
+    )
+
+    parametrize_sriov_invalid_total_vfs = pytest.mark.parametrize(
+        'invalid_values', [-1, 257, 400]
+    )
```
[the remark this answers is not in the document]
Comment (ffmancera) [reply tests/lib/schema_validation_test.py]: Done! Thanks

File: libnmstate/schema.py [hunk 18]
```diff
@@ -200,6 +200,14 @@ class Ethernet(object):
     FULL_DUPLEX = 'full'
     HALF_DUPLEX = 'half'
 
+    SR_IOV = 'sr-iov'
+
+
+class SRIOV(object):
```
[the remark this answers is not in the document]
Comment (ffmancera) [reply libnmstate/schema.py]: We have agreed on that. Done, thanks!

**Comment by ffmancera** [comment]
Rebased, thanks!
